File: notebooksjava/_11MinStack.java
import java.util.ArrayDeque;
import java.util.Deque;

public class _11MinStack {

    // 双スタック協調：_data は全要素、_min のトップは常に現在の最小値
    private final Deque<Integer> data = new ArrayDeque<>();
    private final Deque<Integer> min  = new ArrayDeque<>();

    public void push(int val) {
        data.push(val);
        if (min.isEmpty() || val <= min.peek()) {
            min.push(val);
        }
    }

    public void pop() {
        if (data.isEmpty()) throw new RuntimeException("pop from empty stack");
        int val = data.pop();
        if (val == min.peek()) {
            min.pop();
        }
    }

    public int top() {
        if (data.isEmpty()) throw new RuntimeException("top from empty stack");
        return data.peek();
    }

    public int min() {
        if (min.isEmpty()) throw new RuntimeException("min from empty stack");
        return min.peek();
    }

    public static void main(String[] args) {
        _11MinStack s = new _11MinStack();

        // 基本的な push と min
        s.push(3); s.push(5); s.push(1); s.push(2);
        assert s.min() == 1 : "min should be 1";
        assert s.top() == 2 : "top should be 2";

        // 最小値を pop した後、min が前の値に戻る
        s.pop(); // 2
        s.pop(); // 1（現在の最小値）
        assert s.min() == 3 : "min should restore to 3";
        assert s.top() == 5 : "top should be 5";

        // 重複する最小値
        s.push(3); s.push(3);
        assert s.min() == 3;
        s.pop();
        assert s.min() == 3 : "min should still be 3 after popping one duplicate";

        // 単一要素
        _11MinStack s2 = new _11MinStack();
        s2.push(42);
        assert s2.min() == 42;
        assert s2.top() == 42;

        // 空スタックの操作は例外を投げる
        _11MinStack s3 = new _11MinStack();
        try { s3.pop(); assert false : "should throw"; } catch (RuntimeException e) { /* expected */ }
        try { s3.min(); assert false : "should throw"; } catch (RuntimeException e) { /* expected */ }

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc#/11-min_stack.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

public class MinStack
{
    private Stack<int> data;
    private Stack<int> minStack;

    public MinStack()
    {
        data = new Stack<int>();
        minStack = new Stack<int>();
    }

    // Push value onto stack and maintain min stack
    public void Push(int val)
    {
        data.Push(val);
        if (minStack.Count == 0 || val <= minStack.Peek())
        {
            minStack.Push(val);
        }
    }

    // Pop from stack and update min stack if needed
    public void Pop()
    {
        if (data.Count == 0)
            throw new InvalidOperationException("pop from empty stack");

        int val = data.Pop();
        if (val == minStack.Peek())
        {
            minStack.Pop();
        }
    }

    // Return top element without removing
    public int Top()
    {
        if (data.Count == 0)
            throw new InvalidOperationException("top from empty stack");

        return data.Peek();
    }

    // Return minimum element without removing
    public int Min()
    {
        if (minStack.Count == 0)
            throw new InvalidOperationException("min from empty stack");

        return minStack.Peek();
    }

    static void Main()
    {
        // Test case 1: Basic push and min
        MinStack s = new MinStack();
        s.Push(3);
        s.Push(5);
        s.Push(1);
        s.Push(2);
        Debug.Assert(s.Min() == 1, "min should be 1");
        Debug.Assert(s.Top() == 2, "top should be 2");

        // Test case 2: Min restores after pop
        s.Pop(); // Remove 2
        s.Pop(); // Remove 1 (current min)
        Debug.Assert(s.Min() == 3, "min should restore to 3");
        Debug.Assert(s.Top() == 5, "top should be 5");

        // Test case 3: Duplicate minimum values
        s.Push(3);
        s.Push(3);
        Debug.Assert(s.Min() == 3, "min should be 3");
        s.Pop();
        Debug.Assert(s.Min() == 3, "min should still be 3 after popping one duplicate");

        // Test case 4: Single element
        MinStack s2 = new MinStack();
        s2.Push(42);
        Debug.Assert(s2.Min() == 42, "min should be 42");
        Debug.Assert(s2.Top() == 42, "top should be 42");

        // Test case 5: Pop from empty stack throws exception
        MinStack s3 = new MinStack();
        bool sawPopError = false;
        try
        {
            s3.Pop();
        }
        catch (InvalidOperationException)
        {
            sawPopError = true;
        }
        Debug.Assert(sawPopError, "should throw when popping empty stack");

        // Test case 6: Multiple operations
        MinStack s4 = new MinStack();
        s4.Push(5);
        s4.Push(1);
        s4.Push(3);
        Debug.Assert(s4.Min() == 1);
        s4.Push(1);
        Debug.Assert(s4.Min() == 1);
        s4.Pop();
        Debug.Assert(s4.Min() == 1);
        s4.Pop();
        Debug.Assert(s4.Min() == 1);
        s4.Pop();
        Debug.Assert(s4.Min() == 5);

        Console.WriteLine("All test cases passed ✅");
    }
}
